### PR TITLE
Statemachines

### PR DESCRIPTION
**Pull request by AndreasHeine**
TBC / WIP

### CODE REVIEW

**Comment by AndreasHeine** [comment]
Part 10: Programs
https://reference.opcfoundation.org/v104/Core/docs/Part10/

basic idea:
1. get class instance
2. install a statemachine under a parent node (minimal or fully featured version)
3. use the statemachine api methods to change states, ... 

possible future use: 
-sdk like usable implementation of the basic statemachines
-could be used in instantiate or import xml model to generate functionality from the typedefinition reference / xml model
-for A&C ExclusiveLimitAlarm (which have a ExclusiveLimitStateMachineType) so it could be possible to install a condition and link it to a variable. the transitions of the statemachine also fires the events vor condition change... but for that i need to inspect PR112 for Part 9: Alarms & Conditions

### tbh at this point its experimental till im ready for review!

**Comment by AndreasHeine** [comment]
@oroulet is there a way to tell travis to not run tests on a commit?

**Comment by oroulet** [comment]
Don't think so...

**Comment by swamper123** [comment]
@AndreasHeine You can cancel them either manually (cwhen you click [here](https://travis-ci.org/github/FreeOpcUa/opcua-asyncio/builds/746236414) right side cancel build) or add an `[ci skip]` in your commit message (the internet says)...

**Comment by AndreasHeine** [comment]
UPDATE:
  
the base functionality of the StatemachineType (StateMachineTypeClass) is working, tests and examples will follow!
feel free to leave a comment or give feedback to the API i have tried to keep it intuitive.
  
Statemachine-Test
Statemachine-Events

File: asyncua/common/statemachine.py [hunk 1]
```diff
@@ -0,0 +1,719 @@
...
+            )
+
+    async def init(self, avalible_states, avalible_transitions, ):
+        #FIXME get children and map children
+        #await self.find_all_states()
+        await self.set_avalible_states(avalible_states)
+        #await self.find_all_transitions()
+        await self.set_avalible_transitions(avalible_transitions)
+
+    async def set_avalible_states(self, states):
```
Comment (swamper123) [inline asyncua/common/statemachine.py:284]: A typo: Should be `set_available_states`

Comment (AndreasHeine) [reply asyncua/common/statemachine.py:284]: ty! at least consistently wrong typed ;)

File: asyncua/common/statemachine.py [hunk 2]
```diff
@@ -0,0 +1,719 @@
...
+        '''
+        self._optionals = optionals
+        self._state_machine_node = await self._parent.add_object(
+            self._idx, 
+            self._name, 
+            objecttype=self._state_machine_type, 
+            instantiate_optional=optionals
+            )
+
+    async def init(self, avalible_states, avalible_transitions, ):
```
Comment (swamper123) [inline asyncua/common/statemachine.py:277]: `init `from the parent class `StateMachineTypeClass ` will be overriden and has another signature.

File: asyncua/common/statemachine.py [hunk 3]
```diff
@@ -0,0 +1,719 @@
...
+            self._idx, 
+            self._name, 
+            objecttype=self._state_machine_type, 
+            instantiate_optional=optionals
+            )
+        #FIXME get children and map children
+
+    #Transition
+    async def HaltedToReady(self):
+        await self._current_state.write_value(
```
Comment (swamper123) [inline asyncua/common/statemachine.py:404]: Do you mean `self._current_state_node` ?
I would recommened to delete the _node postfix in the other definitions (unless you got a reason to differ them).

File: asyncua/common/statemachine.py (hunk 2 above)
Comment (AndreasHeine) [reply asyncua/common/statemachine.py:277]: yes init and install must be overriden in the other types but only the base statemachine is woking the rest is WIP maybe i find a better way if i continue with the other statemachine types like finite and programfinite

File: asyncua/common/statemachine.py (hunk 3 above)
Comment (AndreasHeine) [reply asyncua/common/statemachine.py:404]: It is a copy of the VDMA/StateMachineExample/FiniteStateMachine.py we worked earlier on, i just throw it in, will make it consistent thru all types, i promise!

i like to append _node to indicate it is a Node classinstance just for readability but i can also change it afterwards search/replace ;)

**Comment by AndreasHeine** [comment]
just a **keep-alive** message from this PR, iam still working on it! 
Currently reading and structuring my code for substate-machines and substates:
https://reference.opcfoundation.org/v104/Core/docs/Part5/B.5.2/

**Comment by AndreasHeine** [comment]
"StateMachine" and "FiniteStateMachine" are functional and ready to use (examples avalible)
please review and give feedback for the api.
todo:
-improve errorhandling and logging
-maybe add some reference checking in "async init"
-the other spec. defined statemachines (ProgramStateMachine is WIP atm gathering configuration parameters for that type to make it adaptable to the many variants described in spec.)

**Comment by oroulet** [comment]
What is the status of that one? I have no idea what a State machine does in a protocol, but if this ode is necessary and does not touch too many things we can merge it I suppose

**Comment by AndreasHeine** [comment]
the basic StateMachineType and FiniteStatemachineType are ready which are the foundation to the other ones (currently too busy to finish all types in one shot...)

maybe some improvements in errorhandling and on initialistation (ref checking) will come in the future aswell as the missing statemachine types but this can be done in a later PR

**Comment by oroulet** [comment]
OK. We can merge then. But you need to rebase, remove commented code and clean up the history (remove all the merge commit at least) Or I can squash everything in one commit

**Comment by AndreasHeine** [comment]
> Or I can squash everything in one commit

your master ;)

**Comment by AndreasHeine** [comment]
reverting them wasn't a good idea -_-°

**Pull request closed without merging**